File: apps/api/src/modules/knowledge/knowledge.controller.ts
import { Controller, Get, Post, Delete, Param, Body, Patch, UploadedFile, UseInterceptors, Res } from '@nestjs/common';
import { FileInterceptor } from '@nestjs/platform-express';
import { memoryStorage } from 'multer';
import type { Response } from 'express';
import { readFile } from 'node:fs/promises';
import { KnowledgeService } from './knowledge.service.js';
import { KnowledgeRetrievalService } from './knowledge.retrieval.service.js';

@Controller('knowledge')
export class KnowledgeController {
  constructor(
    private readonly knowledgeService: KnowledgeService,
    private readonly retrievalService: KnowledgeRetrievalService,
  ) {
    this.listBases = this.listBases.bind(this);
    this.createBase = this.createBase.bind(this);
    this.updateBase = this.updateBase.bind(this);
    this.deleteBase = this.deleteBase.bind(this);
    this.listBaseDocuments = this.listBaseDocuments.bind(this);
    this.createManualDocument = this.createManualDocument.bind(this);
    this.createUrlDocument = this.createUrlDocument.bind(this);
    this.uploadDocument = this.uploadDocument.bind(this);
    this.getDocument = this.getDocument.bind(this);
    this.getDownloadInfo = this.getDownloadInfo.bind(this);
    this.downloadFile = this.downloadFile.bind(this);
    this.retrieve = this.retrieve.bind(this);
    this.list = this.list.bind(this);
    this.create = this.create.bind(this);
    this.remove = this.remove.bind(this);
    this.listDocuments = this.listDocuments.bind(this);
    this.removeDocument = this.removeDocument.bind(this);
  }

  @Get('bases')
  async listBases() {
    return this.knowledgeService.list();
  }

  @Post('bases')
  async createBase(@Body() body: { name: string; description?: string }) {
    return this.knowledgeService.create(body);
  }

  @Patch('bases/:id')
  async updateBase(@Param('id') id: string, @Body() body: any) {
    return this.knowledgeService.update(id, body);
  }

  @Delete('bases/:id')
  async deleteBase(@Param('id') id: string) {
    return this.knowledgeService.remove(id);
  }

  @Get('bases/:id/documents')
  async listBaseDocuments(@Param('id') id: string) {
    return this.knowledgeService.listDocuments(id);
  }

  @Post('bases/:id/documents/manual')
  async createManualDocument(@Param('id') id: string, @Body() body: { title: string; content: string }) {
    return this.knowledgeService.createManualDocument(id, body);
  }

  @Post('bases/:id/documents/url')
  async createUrlDocument(@Param('id') id: string, @Body() body: { url: string; title?: string }) {
    return this.knowledgeService.createUrlDocument(id, body);
  }

  @Post('bases/:id/documents/upload')
  @UseInterceptors(FileInterceptor('file', { storage: memoryStorage() }))
  async uploadDocument(@Param('id') id: string, @UploadedFile() file: Express.Multer.File) {
    return this.knowledgeService.createUploadedDocument(id, file);
  }

  @Get('documents/:id')
  async getDocument(@Param('id') id: string) {
    return this.knowledgeService.getDocumentById(id);
  }

  @Get('documents/:id/download')
  async getDownloadInfo(@Param('id') id: string) {
    return this.knowledgeService.getDownloadPath(id);
  }

  @Get('documents/:id/file')
  async downloadFile(@Param('id') id: string, @Res() res: Response) {
    const doc: any = await this.knowledgeService.getDocumentById(id);
    if (!doc.filePath) {
      return res.status(404).json({ success: false, error: { code: 'FILE_NOT_FOUND', message: 'No file for document' } });
    }
    const buf = await readFile(doc.filePath);
    res.setHeader('Content-Type', doc.mimeType || 'application/octet-stream');
    res.setHeader('Content-Disposition', `attachment; filename="${encodeURIComponent(doc.fileName || doc.title || 'document')}"`);
    res.send(buf);
  }

  @Post('retrieve')
  async retrieve(@Body() body: { query: string; limit?: number }) {
    return this.retrievalService.retrieve(body.query, body.limit);
  }

  @Get()
  async list() {
    return this.knowledgeService.list();
  }

  @Post()
  async create(@Body() body: { name: string; description?: string }) {
    return this.knowledgeService.create(body);
  }

  @Delete(':id')
  async remove(@Param('id') id: string) {
    return this.knowledgeService.remove(id);
  }

  @Get(':id/documents')
  async listDocuments(@Param('id') id: string) {
    return this.knowledgeService.listDocuments(id);
  }

  @Delete('documents/:id')
  async removeDocument(@Param('id') id: string) {
    return this.knowledgeService.removeDocument(id);
  }
}

File: apps/server-java/src/main/java/com/haoyitec/agent/server/module/knowledge/KnowledgeController.java
package com.haoyitec.agent.server.module.knowledge;

import com.haoyitec.agent.server.common.exception.BizException;
import lombok.RequiredArgsConstructor;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.core.io.Resource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.util.StringUtils;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api/v1/knowledge")
public class KnowledgeController {

    private final KnowledgeService knowledgeService;
    private final KnowledgeRetrievalService retrievalService;

    @GetMapping("/bases")
    public Object listBases() {
        return knowledgeService.list();
    }

    @PostMapping("/bases")
    public Object createBase(@RequestBody Map<String, Object> body) {
        return knowledgeService.create(body);
    }

    @PatchMapping("/bases/{id}")
    public Object updateBase(@PathVariable String id, @RequestBody Map<String, Object> body) {
        return knowledgeService.update(id, body);
    }

    @DeleteMapping("/bases/{id}")
    public Object deleteBase(@PathVariable String id) {
        return knowledgeService.remove(id);
    }

    @GetMapping("/bases/{id}/documents")
    public Object listBaseDocuments(@PathVariable String id) {
        return knowledgeService.listDocuments(id);
    }

    @PostMapping("/bases/{id}/documents/manual")
    public Object createManualDocument(@PathVariable String id, @RequestBody Map<String, Object> body) {
        return knowledgeService.createManualDocument(id, body);
    }

    @PostMapping("/bases/{id}/documents/url")
    public Object createUrlDocument(@PathVariable String id, @RequestBody Map<String, Object> body) {
        return knowledgeService.createUrlDocument(id, body);
    }

    @PostMapping(value = "/bases/{id}/documents/upload", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    public Object uploadDocument(@PathVariable String id, @RequestParam("file") MultipartFile file) {
        return knowledgeService.createUploadedDocument(id, file);
    }

    @GetMapping("/documents/{id}")
    public Object getDocument(@PathVariable String id) {
        return knowledgeService.getDocumentById(id);
    }

    @GetMapping("/documents/{id}/download")
    public Object getDownloadInfo(@PathVariable String id) {
        return knowledgeService.getDownloadPath(id);
    }

    @GetMapping("/documents/{id}/file")
    public ResponseEntity<Resource> downloadFile(@PathVariable String id) {
        @SuppressWarnings("unchecked")
        Map<String, Object> doc = (Map<String, Object>) knowledgeService.getDocumentById(id);
        String filePath = doc.get("filePath") == null ? null : String.valueOf(doc.get("filePath"));
        if (!StringUtils.hasText(filePath)) {
            throw new BizException(HttpStatus.NOT_FOUND, "FILE_NOT_FOUND", "No file for document");
        }
        try {
            byte[] bytes = Files.readAllBytes(Path.of(filePath));
            String fileName = doc.get("fileName") == null ? "document" : String.valueOf(doc.get("fileName"));
            String mimeType = doc.get("mimeType") == null ? "application/octet-stream" : String.valueOf(doc.get("mimeType"));
            return ResponseEntity.ok()
                    .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + fileName + "\"")
                    .contentType(MediaType.parseMediaType(mimeType))
                    .body(new ByteArrayResource(bytes));
        } catch (Exception ex) {
            throw new BizException(HttpStatus.INTERNAL_SERVER_ERROR, "INTERNAL_ERROR", "读取文件失败: " + ex.getMessage());
        }
    }

    @PostMapping("/retrieve")
    public Object retrieve(@RequestBody Map<String, Object> body) {
        String query = body.get("query") == null ? null : String.valueOf(body.get("query"));
        Integer limit = body.get("limit") == null ? null : Integer.parseInt(String.valueOf(body.get("limit")));
        return retrievalService.retrieve(query, limit, null);
    }

    @GetMapping
    public Object list() {
        return knowledgeService.list();
    }

    @PostMapping
    public Object create(@RequestBody Map<String, Object> body) {
        return knowledgeService.create(body);
    }

    @DeleteMapping("/{id}")
    public Object remove(@PathVariable String id) {
        return knowledgeService.remove(id);
    }

    @GetMapping("/{id}/documents")
    public Object listDocuments(@PathVariable String id) {
        return knowledgeService.listDocuments(id);
    }

    @DeleteMapping("/documents/{id}")
    public Object removeDocument(@PathVariable String id) {
        return knowledgeService.removeDocument(id);
    }
}

